### ISSUE TITLE
Ability to publish to S3 buckets with Object Lock by including md5 checksum.

### ISSUE THREAD

**Issue by robsyme** [nextflow-io/nextflow#5347]
## New feature

AWS S3 Object Lock is a feature in Amazon Simple Storage Service (S3) that allows you to protect objects from being deleted or overwritten for a specified retention period. Some orgs will use Object Lock to meet data retention or regulatory requirements.

The `PutObject` call to a bucket with Object Lock requires an extra `Content-MD5 ` header ([docs](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock-managing.html#object-lock-put-object)). Currently, publishing an object to a bucket with Object Lock returns an error:

```
Error executing process > 'MYPROCESS (1)'

Caused by:
  Content-MD5 OR x-amz-checksum- HTTP header is required for Put Object requests with Object Lock parameters (Service: Amazon S3; Status Code: 400; Error Code: InvalidRequest; Request ID: XXXX; S3 Extended Request ID: XXXX; Proxy: null)
```

## Usage scenario 

I'd like to publish a file to a bucket that has [Object Lock](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock.html) turned on.

## Suggest implementation 

For large multipart uploads (often the case with Nextflow published files), the ETag does _not_ represent the MD5 digest, so the object will likely need to be downloaded and hashed by Nextflow so that the `Content-MD5` header can be attached to the publication event.

**Comment by bentsherman** [comment]
What a mess. S3 needs to just accept the ETag for this

**Comment by robsyme** [comment]
100% Agree. It's a little infuriating that a copy from one bucket two another requires a separate digest.

**Comment by zwhitfield** [comment]
@robsyme Sorry to drag up an old thread, but I was curious if you are aware of any workaround or solution for this. I am now facing the same issue.

**Comment by robsyme** [comment]
Hi Zachary. This is a tricky problem to solve at scale. Supporting Object Lock would require Nextflow to download all published files from the task workdir so that we can calculate the file's hash. For large runs with large outputs, this will likely introduce significant performance implications. At the moment, Nextflow does not support publishing to buckets with Object Lock. 
Any change to this will be reflected in this issue.
@bentsherman - given that the Fusion filesystem has the objects in memory anyway, would it be sensible to see if we can have the task's filesystem calculate the hash in situ?

**Comment by bentsherman** [comment]
yes this seems like something that would be best handled by Fusion

**Comment by jorgee** [comment]
I have tested this with 25.10.5, which contains the Java AWS SDK v2. This SDK version adds a default checksum since 2025. Therefore, the error is not raised, and execution succeeds. However, if I am not wrong, Object lock requires enabling versions. This feature prevents deleting or overwriting an object version. If you run a pipeline that generates the same file, it will not fail; it will create another version of this file.

```
$ NXF_VER=25.10.5 nextflow run rnaseq-nf -profile awsbatch -output-dir s3://jorgee-bucket-with-object-lock/work/

 N E X T F L O W   ~  version 25.10.5

Launching `https://github.com/nextflow-io/rnaseq-nf` [modest_ardinghelli] DSL2 - revision: c735b43053 [master]

R N A S E Q - N F   P I P E L I N E
===================================
transcriptome: /home/jorgee/.nextflow/assets/nextflow-io/rnaseq-nf/data/ggal/ggal_1_48850000_49020000.Ggal71.500bpflank.fa
reads        : /home/jorgee/.nextflow/assets/nextflow-io/rnaseq-nf/data/ggal/ggal_gut_{1,2}.fq
outdir       : /jorgee-bucket-with-object-lock/work

Uploading local `bin` scripts folder to s3://jorgee-eu-west1-test1/work/tmp/79/96866b41275cc82f660bb2a395df7a/bin
executor >  awsbatch (4)
[ce/50f352] RNASEQ:INDEX (ggal_1_48850000_49020000) [100%] 1 of 1 ✔
[a2/1297be] RNASEQ:FASTQC (ggal_gut)                [100%] 1 of 1 ✔
[fb/31ed5e] RNASEQ:QUANT (ggal_gut)                 [100%] 1 of 1 ✔
[6d/435323] MULTIQC                                 [100%] 1 of 1 ✔
Completed at: 14-May-2026 16:12:28
Duration    : 3m 31s
CPU hours   : (a few seconds)
Succeeded   : 4
```
